### PR TITLE
fix(modal): native keyboard will dismiss when bottom sheet is dragged

### LINKED ISSUE #23878: bug: sheet modal gesture breaks when interacting with inputs/keyboard on ios
### Prequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [ ] v5.x
- [X] v6.x

### Current Behavior

I just tested the new Sheet Modal. First: Great work 🎉 Here is a Collections of things i mentioned:
- Similar to #23561: My IDEA (InteliJ) seems to not know the new properties like `breakpoints` or `initialBreakpoint`. Maybe missing in the angular component declaration?
- The `handle` can not be set to **false** for Sheet Modals: https://github.com/ionic-team/ionic-framework/blob/9fa373c6efeceb39913f1c158cda383c38cba29c/core/src/components/modal/modal.tsx#L550
- The handle itself doesn't looks nice in combination with IonTitle - Looks like it needs a bit more spacing here 
- The Backdrop seems to be a bit buggy yet, see video: After drag finished it changes:

- When i have an Input inside the Sheet Modal and focus it the Keyboard appears and the Modal pushs up a bit. After losing the focus of the input and the keyboard is gone, the breakpoints do not work anymore and i'm kind of in a free mode where i can stop the Modal wherever i want
- Where we already talk about that: This `freeMode` would be a cool feature
- Also it would be great to be able to disable the possibility to drag the Modal and add the possibility to change the current breakpoint programatically.

### Expected Behavior

Everything is described above

### Steps to Reproduce

Version: `6.0.0-beta.5`

### Code Reproduction URL

_No response_

### Ionic Info

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

The swipe gesture for the modal bottom sheet will trigger when a form element (input, textarea, etc.) receives focus and opens the native keyboard. This causes the sheet gesture to become unresponsive and free scroll. 

Issue Number: Resolves #23878

## What is the new behavior?

- Swipe gestures will be disabled when the keyboard is immediately open (form element has focus).
- Swipe gestures will be re-enabled on the next frame (after the keyboard has repositioned the view)
- The swipe gesture for the sheet modal will dismiss any active keyboards (this aligns with native behavior, see: Maps sheet implementation on iOS)

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

|Before|After|
|---|---|
|<video src="https://user-images.githubusercontent.com/13732623/150849156-0a8ed91f-4c5e-49ea-a1ef-ab88bc31815d.mp4"></video>|<video src="https://user-images.githubusercontent.com/13732623/150849184-5f07511f-1cad-452e-8518-16e61139a7f1.mp4"></video>|

Behavior is easier to feel on device than from a video:
1. Update the sheet example test to render an `<ion-input></ion-input>` for each item
2. Run Core
3. Open the test on your mobile device
4. Present the modal sheet and play with the gesture, breakpoints and keyboard focus

### CODE REVIEW

**Comment by EinfachHans** [comment]
@sean-perkins amazing!! Looks good so far 😊 Dev build would be awesome to test it in a dev app

File: core/src/components/modal/gestures/sheet.ts [hunk 1]
```diff
@@ -42,6 +43,27 @@ export const createSheetGesture = (
...
+       * a single frame and re-enable once the contents have repositioned
+       * from the keyboard placement.
+       */
+      gesture.enable(false);
+      raf(() => {
+        gesture.enable(true)
+      });
+    };
+
+    window.addEventListener(KEYBOARD_DID_OPEN, keyboardOpenCallback);
```
Comment (amandaesmith3) [inline core/src/components/modal/gestures/sheet.ts:64]: It looks like this only gets called once when the modal is presented, but do we need to remove the listener when it's dismissed? 🤔

File: core/src/components/modal/gestures/sheet.ts [hunk 2]
```diff
@@ -42,6 +43,27 @@ export const createSheetGesture = (
   const backdropAnimation = animation.childAnimations.find(ani => ani.id === 'backdropAnimation');
   const maxBreakpoint = breakpoints[breakpoints.length - 1];
 
+  /* tslint:disable-next-line */
```
Comment (amandaesmith3) [inline core/src/components/modal/gestures/sheet.ts:46]: Why the tslint disables? (Here and on L105)

File: core/src/components/modal/gestures/sheet.ts [hunk 3]
```diff
@@ -80,6 +102,15 @@ export const createSheetGesture = (
...
 
+    /* tslint:disable-next-line */
+    if (typeof document !== 'undefined') {
+      const activeElement = baseEl.ownerDocument.activeElement as HTMLElement;
+      if (activeElement.matches('input,ion-input,textarea,ion-textarea')) {
+        // Dismisses the open keyboard when the sheet drag gesture is started.
+        activeElement.blur();
+      }
+    }
+
```
Comment (amandaesmith3) [inline core/src/components/modal/gestures/sheet.ts:113]: Putting this in `canStart` feels a little weird; it seems like it should go in `onStart` or somewhere else that more explicitly indicates that we're starting a gesture.

File: core/src/components/modal/gestures/sheet.ts (hunk 2 above)
Comment (sean-perkins) [reply core/src/components/modal/gestures/sheet.ts:46]: The linter errors out that the check is always true, but that isn't the case in SSR. We use this approach in almost all cases where we are checking if the document or window is available.

File: core/src/components/modal/gestures/sheet.ts (hunk 1 above)
Comment (sean-perkins) [reply core/src/components/modal/gestures/sheet.ts:64]: That's a good point. I had similar concessions, but wasn't able to find anywhere in gestures that we expose appending additional behavior to teardown. Let me see if there's a way to extend the `destroy` method of the gesture (which is called in the dismiss callback from the modal).

File: core/src/components/modal/gestures/sheet.ts (hunk 2 above)
Comment (amandaesmith3) [reply core/src/components/modal/gestures/sheet.ts:46]: Ah, gotcha 👍 I wasn't getting an error on my end after removing the disable comment, so I wasn't sure. I'm seeing it in other places though, like you said.

File: core/src/components/modal/gestures/sheet.ts (hunk 3 above)
Comment (sean-perkins) [reply core/src/components/modal/gestures/sheet.ts:113]: Agreed 🤔  let me confirm that doing this behavior while the gesture is starting, doesn't cause any other issues. Then I'll move it 👍

**Review by amandaesmith3 — APPROVED** [review]
Looks great, nice work 👍

**Comment by sean-perkins** [comment]
> @sean-perkins amazing!! Looks good so far 😊 Dev build would be awesome to test it in a dev app

@EinfachHans The dev build version is `6.0.4-dev.1643056868.324d279`.

i.e.:
```bash
npm i @ionic/angular@6.0.4-dev.1643056868.324d279
```

**Review by EinfachHans — APPROVED** [review]
@sean-perkins works as expected! 😊

File: core/src/components/modal/gestures/sheet.ts [hunk 4]
```diff
@@ -93,6 +115,17 @@ export const createSheetGesture = (
       contentEl.scrollY = false;
     }
 
+    /* tslint:disable-next-line */
+    if (typeof document !== 'undefined') {
+      const activeElement = baseEl.ownerDocument.activeElement as HTMLElement;
+      if (activeElement.matches('input,ion-input,textarea,ion-textarea')) {
```
[the remark this answers is not in the document]
Comment (ccapndave) [reply core/src/components/modal/gestures/sheet.ts:121]: I'm using `contenteditable` (with Lexical) and this does indeed break!

**Pull request merged**